pkg/collector/harbor_events.go

```diff
@@ -0,0 +1,508 @@
+package collector
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/go-logr/logr"
+	discovery "github.com/grafeas/grafeas/proto/v1beta1/discovery_go_proto"
+	grafeas "github.com/grafeas/grafeas/proto/v1beta1/grafeas_go_proto"
+	image "github.com/grafeas/grafeas/proto/v1beta1/image_go_proto"
+	packag "github.com/grafeas/grafeas/proto/v1beta1/package_go_proto"
+	vulnerability "github.com/grafeas/grafeas/proto/v1beta1/vulnerability_go_proto"
+	"github.com/liatrio/rode/pkg/occurrence"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type HarborEventCollector struct {
+	logger    logr.Logger
+	url       string
+	secret    string
+	project   string
+	namespace string
+}
+
+func NewHarborEventCollector(logger logr.Logger, harborURL string, secret string, project string, namespace string) Collector {
+	return &HarborEventCollector{
+		logger:    logger,
+		url:       harborURL,
+		secret:    secret,
+		project:   project,
+		namespace: namespace,
+	}
+}
+
+func (t *HarborEventCollector) Reconcile(ctx context.Context, name types.NamespacedName) error {
+	harborCreds, err := t.getHarborCredentials(ctx, t.secret, t.namespace)
+	if err != nil {
+		return err
+	}
+
+	projectID, err := t.getProjectID(t.project, t.url)
+	if err != nil {
+		return err
+	}
+	webhookCheck, err := t.checkForWebhook(projectID, t.url, harborCreds)
+	if err != nil {
+		return err
+	}
+	if !webhookCheck {
+		//TODO: Assuming ingress is called rode deployed into rode namespace
+		hostName, err := t.getHostName("rode", "rode")
+		if err != nil {
+			return err
+		}
+		err = t.createWebhook(projectID, t.url, harborCreds, hostName+"/webhook/harbor_event/"+name.String())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *HarborEventCollector) Destroy(ctx context.Context) error {
+	t.logger.Info("destroying test collector")
+	harborCreds, err := t.getHarborCredentials(ctx, t.secret, t.namespace)
+	if err != nil {
+		return err
+	}
+	projectID, err := t.getProjectID(t.project, t.url)
+	if err != nil {
+		return err
+	}
+	policyID, err := t.getWebhookPolicyID(projectID, t.url, harborCreds)
+	if err != nil {
+		return err
+	}
+	err = t.deleteWebhookPolicy(projectID, t.url, policyID, harborCreds)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *HarborEventCollector) Type() string {
+	return "harbor_event"
+}
+
+func (t *HarborEventCollector) HandleWebhook(writer http.ResponseWriter, request *http.Request, occurrenceCreator occurrence.Creator) {
+	var payload *Payload
+	body, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		log.Fatal(err)
+		writer.WriteHeader(http.StatusInternalServerError)
```

Comment:
If a body was not provided, that should be a bad request (400), not a server error (500).